nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/common/AbstractMQTTProcessor.java

```diff
@@ -0,0 +1,342 @@
+
+
+    public static final PropertyDescriptor PROP_CLIENTID = new PropertyDescriptor.Builder()
+            .name("Client ID")
+            .description("MQTT client ID to use")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_USERNAME = new PropertyDescriptor.Builder()
+            .name("Username")
+            .description("Username to use when connecting to the broker")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_PASSWORD = new PropertyDescriptor.Builder()
+            .name("Password")
+            .description("Password to use when connecting to the broker")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_SSL_CONTEXT_SERVICE = new PropertyDescriptor.Builder().name("SSL Context Service")
+            .description("The SSL Context Service used to provide client certificate information for TLS/SSL connections.")
+            .required(false)
+            .identifiesControllerService(SSLContextService.class)
+            .build();
+
+
+    public static final PropertyDescriptor PROP_LAST_WILL_TOPIC = new PropertyDescriptor.Builder()
+            .name("Last Will Topic")
+            .description("The topic to send the client's Last Will to. If the Last Will topic and message are not set then a Last Will will not be sent.")
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_LAST_WILL_MESSAGE = new PropertyDescriptor.Builder()
+            .name("Last Will Message")
+            .description("The message to send as the client's Last Will. If the Last Will topic and message are not set then a Last Will will not be sent.")
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_LAST_WILL_RETAIN = new PropertyDescriptor.Builder()
+            .name("Last Will Retain")
+            .description("Whether to retain the client's Last Will. If the Last Will topic and message are not set then a Last Will will not be sent.")
+            .required(false)
+            .allowableValues("true","false")
+            .build();
+
+    public static final PropertyDescriptor PROP_LAST_WILL_QOS = new PropertyDescriptor.Builder()
+            .name("'Last Will' QoS Level")
+            .description("QoS level to be used when publishing the Will Message")
+            .required(false)
+            .allowableValues(
+                    ALLOWABLE_VALUE_QOS_0,
+                    ALLOWABLE_VALUE_QOS_1,
+                    ALLOWABLE_VALUE_QOS_2
+            )
```

Comment:
This was intentional, because some applications may not have a Last Will and there is no point forcing a use to have it set when it wouldn't be used. 
That said I should have better validation to make sure it is set if the other Last Will properties are.